File: backend/app/alerts.py
"""Alert Service and Resolution Subsystem for VoxGuard.

Implements the contract in docs/02-API_CONTRACTS.md §10 & §10.1 (DR-008, DR-019):
- Manages deduplicated, idempotency-keyed alerts (one per band escalation).
- Enforces Human-in-the-loop SLA tracking (Ack & Resolve deadlines).
- Supports resolution outcomes: CONFIRMED_FRAUD, FALSE_POSITIVE, INCONCLUSIVE.
- Integrates customer appeal filing to guarantee compliance and fairness.
"""
from datetime import datetime, timezone, timedelta
from uuid import uuid4
import logging

logger = logging.getLogger(__name__)


def calculate_sla_deadlines(band: str, created_at_dt: datetime) -> tuple[str, str]:
    """Computes SLA Ack & Resolve timestamps according to docs/02 §10.1."""
    if band == "HIGH":
        ack_dt = created_at_dt + timedelta(minutes=2)
        resolve_dt = created_at_dt + timedelta(minutes=30)
    elif band == "MEDIUM":
        ack_dt = created_at_dt + timedelta(minutes=15)
        resolve_dt = created_at_dt + timedelta(hours=4)
    else:
        ack_dt = created_at_dt + timedelta(hours=1)
        resolve_dt = created_at_dt + timedelta(hours=24)
    return ack_dt.isoformat(), resolve_dt.isoformat()


def make_alert(call_id: str, score: float, band: str, alert_id: str) -> dict:
    """Creates a new structured alert record with SLA tracking."""
    now_dt = datetime.now(timezone.utc)
    ack_sla, resolve_sla = calculate_sla_deadlines(band, now_dt)
    
    logger.info("Security Alert created for call %s: band=%s score=%.2f id=%s", call_id, band, score, alert_id)
    
    return {
        "id": alert_id,
        "call_id": call_id,
        "risk_score": score,
        "band": band,
        "status": "active",
        "created_at": now_dt.isoformat(),
        "updated_at": now_dt.isoformat(),
        "sla_ack_deadline": ack_sla,
        "sla_resolve_deadline": resolve_sla,
        "assigned_to": None,
        "resolution": None,
        "appeals": [],
        "message": "Additional verification required. Voice synthesis and contextual anomalies detected.",
        "recommendations": [
            "Out-of-band callback using verified profile number",
            "Require in-app biometric / hardware token MFA",
            "Hold high-value transactions pending supervisor sign-off"
        ],
        "auto_block": band == "HIGH",
        "notification_channel": "DashboardChannel (with Webhook retry lane)"
    }


def assign_alert(alert: dict, assignee_id: str) -> dict:
    """Assigns an alert to an investigator, moving it to 'assigned' status."""
    alert["assigned_to"] = assignee_id
    if alert["status"] == "active":
        alert["status"] = "assigned"
    alert["updated_at"] = datetime.now(timezone.utc).isoformat()
    return alert


def resolve_alert(alert: dict, outcome: str, notes: str, resolver_id: str) -> dict:
    """Resolves an alert with a mandatory closed-enum outcome for feedback loop training."""
    valid_outcomes = {"CONFIRMED_FRAUD", "FALSE_POSITIVE", "INCONCLUSIVE"}
    if outcome not in valid_outcomes:
        raise ValueError(f"Invalid resolution outcome '{outcome}'. Must be one of {valid_outcomes}")
    if not notes or len(notes.strip()) < 3:
        raise ValueError("Investigation notes are mandatory when resolving a security alert.")

    alert["status"] = "resolved"
    alert["updated_at"] = datetime.now(timezone.utc).isoformat()
    alert["resolution"] = {
        "outcome": outcome,
        "notes": notes,
        "resolver_id": resolver_id,
        "resolved_at": datetime.now(timezone.utc).isoformat(),
    }
    return alert


def lodge_appeal(alert: dict, reason: str, appellant_type: str = "CUSTOMER", contact_info: str | None = None) -> dict:
    """Lodges an appeal ticket against an alert or stepped-up decision per docs/09 §5."""
    if not reason or len(reason.strip()) < 5:
        raise ValueError("A detailed explanation is required to lodge an appeal.")
    
    appeal_ticket = {
        "appeal_id": f"app-{uuid4().hex[:8]}",
        "alert_id": alert["id"],
        "call_id": alert["call_id"],
        "appellant_type": appellant_type,
        "reason": reason,
        "contact_info": contact_info or "Not provided",
        "status": "PENDING_REVIEW",
        "sla_resolution_deadline": (datetime.now(timezone.utc) + timedelta(hours=24)).isoformat(),
        "filed_at": datetime.now(timezone.utc).isoformat(),
    }
    alert["appeals"].append(appeal_ticket)
    alert["updated_at"] = datetime.now(timezone.utc).isoformat()
    return appeal_ticket

File: gateway/internal/alerts/alerts.go
// Package alerts is the Alert Service (docs/01 section 2, docs/02 sections 10 and 10.1,
// DR-008, DR-019, [Go]).
//
// It holds deduplicated, idempotency-keyed alerts — one per band escalation, never one per
// window (CLAUDE.md invariant 11) — tracks the human-in-the-loop SLA, and records resolution
// outcomes and customer appeals.
package alerts

import (
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ErrInvalidOutcome reports a resolution outcome outside the closed enum. The enum is closed
// because these outcomes are the feedback signal a future detector is trained on; free text
// would make the loop unusable.
var ErrInvalidOutcome = errors.New("invalid resolution outcome")

// ErrNotesRequired reports a resolution with no investigation notes.
var ErrNotesRequired = errors.New("investigation notes are mandatory when resolving an alert")

// ErrReasonRequired reports an appeal with no stated reason.
var ErrReasonRequired = errors.New("a detailed explanation is required to lodge an appeal")

// ErrNotFound reports an unknown alert id.
var ErrNotFound = errors.New("alert not found")

var validOutcomes = map[string]bool{
	"CONFIRMED_FRAUD": true, "FALSE_POSITIVE": true, "INCONCLUSIVE": true,
}

// Resolution closes an alert with an outcome an analyst stands behind.
type Resolution struct {
	Outcome    string `json:"outcome"`
	Notes      string `json:"notes"`
	ResolverID string `json:"resolver_id"`
	ResolvedAt string `json:"resolved_at"`
}

// Appeal is a customer or agent challenge to an alert or a stepped-up decision (docs/09 §5).
type Appeal struct {
	AppealID              string `json:"appeal_id"`
	AlertID               string `json:"alert_id"`
	CallID                string `json:"call_id"`
	AppellantType         string `json:"appellant_type"`
	Reason                string `json:"reason"`
	ContactInfo           string `json:"contact_info"`
	Status                string `json:"status"`
	SLAResolutionDeadline string `json:"sla_resolution_deadline"`
	FiledAt               string `json:"filed_at"`
}

// Alert is one escalation raised to the fraud desk.
type Alert struct {
	ID        string  `json:"id"`
	CallID    string  `json:"call_id"`
	RiskScore float64 `json:"risk_score"`
	Band      string  `json:"band"`
	Status    string  `json:"status"`

	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`

	SLAAckDeadline     string `json:"sla_ack_deadline"`
	SLAResolveDeadline string `json:"sla_resolve_deadline"`

	AssignedTo *string     `json:"assigned_to"`
	Resolution *Resolution `json:"resolution"`
	Appeals    []Appeal    `json:"appeals"`

	// Message and Recommendations are read by an agent mid-call. CLAUDE.md invariant 12: they
	// say what verification is required, never that the person on the line is a fraudster.
	Message         string   `json:"message"`
	Recommendations []string `json:"recommendations"`

	AutoBlock           bool   `json:"auto_block"`
	NotificationChannel string `json:"notification_channel"`

	PolicyVersion string            `json:"policy_version"`
	ModelVersions map[string]string `json:"model_versions"`
}

// SLADeadlines computes the acknowledge and resolve deadlines for a band (docs/02 §10.1).
func SLADeadlines(band string, createdAt time.Time) (string, string) {
	var ack, resolve time.Duration
	switch band {
	case "HIGH":
		ack, resolve = 2*time.Minute, 30*time.Minute
	case "MEDIUM":
		ack, resolve = 15*time.Minute, 4*time.Hour
	default:
		ack, resolve = time.Hour, 24*time.Hour
	}
	return createdAt.Add(ack).Format(time.RFC3339Nano), createdAt.Add(resolve).Format(time.RFC3339Nano)
}

// Store holds the alerts for this process.
type Store struct {
	mu      sync.RWMutex
	byID    map[string]*Alert
	ordered []*Alert
}

// NewStore builds an empty alert store.
func NewStore() *Store {
	return &Store{byID: map[string]*Alert{}}
}

// Ensure returns the alert for an idempotency key, creating it on first sight and refreshing the
// score on every later sight. The key is the escalation key from the session aggregator, so a
// session that stays HIGH for hundreds of windows still yields exactly one alert per escalation.
// Created reports whether this call was the one that raised it.
func (s *Store) Ensure(key, callID string, score float64, band, policyVersion string,
	modelVersions map[string]string) (alert *Alert, created bool) {

	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now().UTC()
	if existing, ok := s.byID[key]; ok {
		existing.RiskScore = score
		existing.UpdatedAt = now.Format(time.RFC3339Nano)
		return existing, false
	}

	ack, resolve := SLADeadlines(band, now)
	a := &Alert{
		ID:                 key,
		CallID:             callID,
		RiskScore:          score,
		Band:               band,
		Status:             "active",
		CreatedAt:          now.Format(time.RFC3339Nano),
		UpdatedAt:          now.Format(time.RFC3339Nano),
		SLAAckDeadline:     ack,
		SLAResolveDeadline: resolve,
		Appeals:            []Appeal{},
		Message:            "Additional verification required. Voice synthesis and contextual anomalies detected.",
		Recommendations: []string{
			"Out-of-band callback using verified profile number",
			"Require in-app biometric / hardware token MFA",
			"Hold high-value transactions pending supervisor sign-off",
		},
		AutoBlock:           band == "HIGH",
		NotificationChannel: "DashboardChannel (with Webhook retry lane)",
		PolicyVersion:       policyVersion,
		ModelVersions:       modelVersions,
	}
	s.byID[key] = a
	s.ordered = append(s.ordered, a)
	return a, true
}

// Get returns one alert by id.
func (s *Store) Get(id string) (*Alert, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	a, ok := s.byID[id]
	if !ok {
		return nil, ErrNotFound
	}
	return a, nil
}

// List returns every alert, newest first.
func (s *Store) List() []*Alert {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]*Alert, 0, len(s.ordered))
	for i := len(s.ordered) - 1; i >= 0; i-- {
		out = append(out, s.ordered[i])
	}
	return out
}

// Assign routes an alert to an investigator.
func (s *Store) Assign(id, assigneeID string) (*Alert, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.byID[id]
	if !ok {
		return nil, ErrNotFound
	}
	a.AssignedTo = &assigneeID
	if a.Status == "active" {
		a.Status = "assigned"
	}
	a.UpdatedAt = time.Now().UTC().Format(time.RFC3339Nano)
	return a, nil
}

// Resolve closes an alert with a mandatory closed-enum outcome and notes.
func (s *Store) Resolve(id, outcome, notes, resolverID string) (*Alert, error) {
	if !validOutcomes[outcome] {
		return nil, fmt.Errorf("%w %q: must be CONFIRMED_FRAUD, FALSE_POSITIVE or INCONCLUSIVE",
			ErrInvalidOutcome, outcome)
	}
	if len(strings.TrimSpace(notes)) < 3 {
		return nil, ErrNotesRequired
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.byID[id]
	if !ok {
		return nil, ErrNotFound
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	a.Status = "resolved"
	a.UpdatedAt = now
	a.Resolution = &Resolution{Outcome: outcome, Notes: notes, ResolverID: resolverID, ResolvedAt: now}
	return a, nil
}

// Escalate marks an alert escalated. It reports whether this call was the transition, so the
// caller writes exactly one audit record rather than one per click.
func (s *Store) Escalate(id string) (*Alert, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.byID[id]
	if !ok {
		return nil, false, ErrNotFound
	}
	if a.Status == "escalated" {
		return a, false, nil
	}
	a.Status = "escalated"
	a.UpdatedAt = time.Now().UTC().Format(time.RFC3339Nano)
	return a, true, nil
}

// LodgeAppeal files a challenge against an alert (docs/09 §5).
//
// The appeal path is not optional politeness. A system that can require step-up verification on
// the strength of an unvalidated acoustic model owes the person on the other end a route to
// contest it.
func (s *Store) LodgeAppeal(id, reason, appellantType, contactInfo string) (*Appeal, error) {
	if len(strings.TrimSpace(reason)) < 5 {
		return nil, ErrReasonRequired
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.byID[id]
	if !ok {
		return nil, ErrNotFound
	}
	if contactInfo == "" {
		contactInfo = "Not provided"
	}
	if appellantType == "" {
		appellantType = "CUSTOMER"
	}
	now := time.Now().UTC()
	appeal := Appeal{
		AppealID:              "app-" + uuid.NewString()[:8],
		AlertID:               a.ID,
		CallID:                a.CallID,
		AppellantType:         appellantType,
		Reason:                reason,
		ContactInfo:           contactInfo,
		Status:                "PENDING_REVIEW",
		SLAResolutionDeadline: now.Add(24 * time.Hour).Format(time.RFC3339Nano),
		FiledAt:               now.Format(time.RFC3339Nano),
	}
	a.Appeals = append(a.Appeals, appeal)
	a.UpdatedAt = now.Format(time.RFC3339Nano)
	return &appeal, nil
}
